### ISSUE TITLE
v6: relative field names i.e. inherit parent name path

### ISSUE THREAD

**Issue by jayphelps** [erikras/redux-form#1132]
##### tl;dr

Could redux-form support an elegant, first-class way to have field names be relative to the parent field within a form?

``` jsx
// first.second
<Field
  name="first"
  component={field =>
    // have to specify fully-qualified path (is not relative)
    <Field name="first.second" component="input" />
  }
/>
```

The obvious solution would be to make them always relative, but that means you then might need a way to define a name that _shouldn't_ be relative. Another possibility is a flag `<Field name="foo" relative />`. Perhaps you can think of others.
##### Explaination

I deal with rather complex, multilayer forms. With the new v6 API, I find myself needing to pass around field names all over the place so that the form data takes the proper structure without having the fully qualified name path being hardcoded; since these fields get used in multiple forms in varying levels.

Here's just a taste of one example, which isn't even the most complex we have and is stripped down for additional clarity:

Possible fields:

``` jsx
'sources[].name',
'sources[].expression',
'sources[].metadata.fields[]',
'sources[].metadata.combineOperator',
'sources[].metadata.conditions[].name',
'sources[].metadata.conditions[].operator',
'sources[].metadata.conditions[].type',
'sources[].metadata.conditions[].value'
```

And here's a possible component structure to build some of this, with most of the layout-related elements stripped out since they're extraneous for this example. Some of the fields are excluded since they're fairly obvious what they would contain.

``` jsx
// We use this to possibly inherit the field name from the parent.
// Some times the component will *not* be inside another field,
// but sometimes they will be, so this is handy for the component
// to not need to know
const concatPaths = (prefix, path) => prefix ? prefix + '.' + path : path;

/********************************************/
/* Start at the bottom and work your way up */
/********************************************/

const ConditionsFieldArray = ({ fields }) => (
  <div>
    {fields.map((fieldPath, index) =>
      <Condition
        key={index}
        fieldPath={fieldPath}
        remove={() => fields.remove(index)}
      />
    )}
    <button
      type="button"
      className="ui button"
      onClick={() => fields.push(createCondition())}
    >
      <i className="plus icon" />
      Add condition
    </button>
  </div>
);

const Conditions = ({ fieldPath }) => (
  <FieldArray
    name={concatPaths(fieldPath, 'conditions')}
    component={ConditionsFieldArray}
  />
);

const MetadataField = ({ name: fieldPath }) => (
  <div>
    <div>
      <Fields fieldPath={fieldPath} />
    </div>
    <div>
      <Conditions fieldPath={fieldPath} />
    </div>
    <div>
      <CombineOperatorField fieldPath={fieldPath} />
    </div>
  </div>
);

const Metadata = ({ fieldPath }) => (
  <Field
    name={concatPaths(fieldPath, 'metadata')}
    component={MetadataField}
  />
);

const Source = ({ fieldPath }) => (
  <div>
    <Name fieldPath={fieldPath} />
    <Expression fieldPath={fieldPath} />
    <Metadata fieldPath={fieldPath} />
  </div>
);

const SourcesFieldArray = ({ fields }) => (
  <div>
    {fields.map((fieldPath, index) =>
      <Source key={index} fieldPath={fieldPath} />
    )}
  </div>
);

const Sources = ({ fieldPath }) => (
  <FieldArray
    name={concatPaths(fieldPath, 'sources')}
    component={SourcesFieldArray}
  />
);

export default reduxForm({ form: 'query-wizard' });
```

With this stripped down example you may ask why we are extracting fields into numerous components. Excluded from this example are the additional complexities with these fields needing to connect to the store to get options and dispatch other effects, which means in the end if we did this all in one or two components they'd be massively monolithic. We also reuse these components in numerous places, with differing form state structures. Perhaps we're just a very rare use case? The counter to that argument is that simple use cases less justify using an abstraction like redux-form IMO.

On an unrelated note, the need to create wrapper components has proven to be rather annoying in practice. Although I understand why it was chosen, it definitely has made things feel verbose and makes depending and changing other field values based on your own field changes, more tough. We have a ton of other feedback we've experienced so far trying to port from v5 to v6, let me know if there's an ideal forum to discuss this other than making tickets for everything. e.g. it would make our lives easier if form name was included on the `field` descriptor `({ form, name, value, onChange }) => ...` and whether `component="input"` should be default of `Field`

**Comment by ooflorent** [comment]
When we discussed the v6 API I suggested something that wasn't accepted: instead of using `name="path.to.field"` you could use an object holding the path.

```js
<FieldArray name={fields.sources} />
<Field name={fields.deep.field} />
```

It could be easily typed:

```js
union Field = FieldMap | FieldList | FieldValue

declare class FieldNode {
  path: string;
}

declare class FieldMap extends FieldNode {
  type: "FieldMap";
  [key: string]: Field;
}

declare class FieldList extends FieldNode {
  type: "FieldList";
  length: number;
  [key: number]: Field;
}

declare class FieldValue extends FieldNode {
  type: "FieldValue";
}
```

Here is a sample tree:

```js
{
  "sources": {
    "@@type": "FieldList",
    "@@path": "sources",
    "length": 1,
    "0": {
      "@@type": "FieldMap",
      "@@path": "sources[0]",
      "name": {
        "@@type": "FieldValue",
        "@@path": "sources[0].name"
      },
      "expression": {
        "@@type": "FieldValue",
        "@@path": "sources[0].expression"
      },
      "metadata": {
        "@@type": "FieldMap",
        "@@path": "sources[0].metadata",
        "fields": {
          "@@type": "FieldList",
          "@@path": "sources[0].metadata.fields",
          "length": 0
        }
      }
    }
  }
}
```

Such API would solve your issue.
cc @erikras

**Comment by jayphelps** [comment]
One other reason I need to pass the field name is because it's the only way (I'm aware of) for fields inside of an array to know what index they're at. That's something you can't hardcode.

**Comment by erikras** [comment]
> When we discussed the `v6` API I suggested something that wasn't accepted: instead of using `name="path.to.field"` you could use an object holding the path.

@jayphelps's deep structure is an excellent example of why such a `fields` structure, like that of `v5`, would be extremely helpful. A few things:

Mutating that `this.props.fields` structure properly was an endless source of bugs in `v5` (there are still some outstanding ones), and I have very much enjoyed its lack of existence when working on `v6`. 

That said, I have always planned on eventually giving it another try. One of the first things I did when writing `v6` (even before the reducer!) was to write a leaf-agnostic [`generateFields()`](https://github.com/erikras/redux-form/blob/544815fb784afa652120b079b352fc30a38a7cd0/src/generateFields.js) function to generate that structure from a list of field name strings. But I have been putting off attempting this until `v6` is stable and released.

In the meantime...

@jayphelps, what are you gaining from `first` being a `Field` in your `first.second` example? But yes, passing around the `fieldPath` is the only way to do what you want at the moment. `FieldArray`, of course, provides you with the path that came before.

@ooflorent What is this typing language you are using?

**Comment by ooflorent** [comment]
@jayphelps With the above proposed syntax, passing field names is not an issue! Since `name` no longer receive a string but a path descriptor containing its children, you can the current path.

@erikras It's Flow's declare syntax.

**Comment by ooflorent** [comment]
@erikras In my proposal, `fields` only holds the structure, not the values neither the flags (`dirty`, `touched`, etc) so it won't trigger excessive renders and is mutated only when the field structure changes.

**Comment by jayphelps** [comment]
@erikras 

> what are you gaining from first being a Field in your first.second example?

That was a somewhat contrived example, but in my particular instance it's simply because the API expects all the inner fields to be nested under that object. Keeping the UI in-sync with the API makes things much easier to grok in such a complex form and it prevents us from having naming conflicts later, which has happened when we "lifted" fields because _at the time_ they were unique.

**Comment by huan086** [comment]
Does the new FormSection sufficient to solve this?

**Comment by gustavohenke** [comment]
What @huan086 said 👆, use the `FormSection` componenet.
